File: mobile/ios/Runner/Services/fiat/FiatRampService.swift
import Foundation

class FiatRampService {
    static let API_BASE = "https://api.tigerwallet.com/api/v1"
    var authToken: String?
    
    init(token: String?) { self.authToken = token }
    
    private var headers: [String: String] {
        var h = ["Content-Type": "application/json"]
        if let token = authToken { h["Authorization"] = "Bearer \(token)" }
        return h
    }
    
    func buyCrypto(amount: Double, currency: String, crypto: String, completion: @escaping (String?) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/fiat/buy") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["amount": amount, "currency": currency, "crypto": crypto])
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let orderId = (json["data"] as? [String: Any])?["orderId"] as? String else {
                completion(nil)
                return
            }
            completion(orderId)
        }.resume()
    }
    
    func sellCrypto(amount: Double, crypto: String, currency: String, completion: @escaping (String?) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/fiat/sell") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["amount": amount, "crypto": crypto, "currency": currency])
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let orderId = (json["data"] as? [String: Any])?["orderId"] as? String else {
                completion(nil)
                return
            }
            completion(orderId)
        }.resume()
    }
}

class GiftCardService {
    static let API_BASE = "https://api.tigerwallet.com/api/v1"
    var authToken: String?
    
    init(token: String?) { self.authToken = token }
    
    private var headers: [String: String] {
        var h = ["Content-Type": "application/json"]
        if let token = authToken { h["Authorization"] = "Bearer \(token)" }
        return h
    }
    
    func getBrands(completion: @escaping ([GiftCardBrand]) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/giftcards/brands") else { return }
        var request = URLRequest(url: url)
        request.allHTTPHeaderFields = headers
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let array = json["data"] as? [[String: Any]] else {
                completion([])
                return
            }
            let brands = array.compactMap { dict -> GiftCardBrand? in
                guard let id = dict["id"] as? String, let name = dict["name"] as? String else { return nil }
                return GiftCardBrand(id: id, name: name, discount: dict["discount"] as? Double ?? 0)
            }
            completion(brands)
        }.resume()
    }
    
    func buyGiftCard(brandId: String, amount: Double, completion: @escaping (String?) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/giftcards/buy") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["brandId": brandId, "amount": amount])
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let code = (json["data"] as? [String: Any])?["code"] as? String else {
                completion(nil)
                return
            }
            completion(code)
        }.resume()
    }
}

struct GiftCardBrand { let id: String; let name: String; let discount: Double }

class DAppBrowserService {
    static let API_BASE = "https://api.tigerwallet.com/api/v1"
    var authToken: String?
    
    init(token: String?) { self.authToken = token }
    
    private var headers: [String: String] {
        var h = ["Content-Type": "application/json"]
        if let token = authToken { h["Authorization"] = "Bearer \(token)" }
        return h
    }
    
    func getFeaturedDApps(completion: @escaping ([DApp]) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/dapp/featured") else { return }
        var request = URLRequest(url: url)
        request.allHTTPHeaderFields = headers
        
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let array = json["data"] as? [[String: Any]] else {
                completion([])
                return
            }
            let dapps = array.compactMap { dict -> DApp? in
                guard let id = dict["id"] as? String, let name = dict["name"] as? String, let url = dict["url"] as? String else { return nil }
                return DApp(id: id, name: name, url: url, category: dict["category"] as? String ?? "")
            }
            completion(dapps)
        }.resume()
    }
    
    func connectWallet(dappId: String, address: String, chain: String, completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: "\(Self.API_BASE)/dapp/connect") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.httpBody = try? JSONSerialization.data(withJSONObject: ["dappId": dappId, "address": address, "chain": chain])
        
        URLSession.shared.dataTask(with: request) { _, response, _ in
            if let http = response as? HTTPURLResponse { completion(http.statusCode == 201) }
            else { completion(false) }
        }.resume()
    }
}

struct DApp { let id: String; let name: String; let url: String; let category: String }

File: mobile_apps/android_legacy/app/src/main/java/com/tigerwallet/app/services/fiat/FiatRampService.java
package com.tigerwallet.app.services;

import org.json.JSONObject;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;

public class FiatRampService {
    private static final String API_BASE = "https://api.tigerwallet.com/api/v1";
    private String authToken;
    
    public FiatRampService(String token) { this.authToken = token; }
    
    public String buyCrypto(String amount, String currency, String crypto) throws Exception {
        URL url = new URL(API_BASE + "/fiat/buy");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        conn.setDoOutput(true);
        JSONObject body = new JSONObject();
        body.put("amount", amount);
        body.put("currency", currency);
        body.put("crypto", crypto);
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        if (conn.getResponseCode() == 201) {
            String response = readResponse(conn);
            return new JSONObject(response).getJSONObject("data").getString("orderId");
        }
        return null;
    }
    
    public String sellCrypto(double amount, String crypto, String currency) throws Exception {
        URL url = new URL(API_BASE + "/fiat/sell");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        conn.setDoOutput(true);
        JSONObject body = new JSONObject();
        body.put("amount", amount);
        body.put("crypto", crypto);
        body.put("currency", currency);
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        if (conn.getResponseCode() == 201) {
            String response = readResponse(conn);
            return new JSONObject(response).getJSONObject("data").getString("orderId");
        }
        return null;
    }
    
    public String getStatus(String orderId) throws Exception {
        URL url = new URL(API_BASE + "/fiat/order/" + orderId);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            return new JSONObject(response).getJSONObject("data").getString("status");
        }
        return null;
    }
    
    private String readResponse(HttpURLConnection conn) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) response.append(line);
        reader.close();
        return response.toString();
    }
}
